File: platform/polyglot/app_yaml_overwrites/mjs/src/context-builder.ts

import { create, ILogger } from './logger.js';
import { SharedContext, createSharedContext } from './shared-context.js';

/**
 * Generic request type for framework-agnostic usage.
 */
export interface RequestLike {
    headers?: Record<string, string | string[] | undefined>;
    [key: string]: any;
}

export interface ContextOptions {
    env?: Record<string, string>;
    config?: any; // AppYamlConfig raw
    app?: any;    // Extracted app config
    state?: any;
    request?: RequestLike;
    shared?: SharedContext;
}

export type ContextExtender = (currentContext: any, request?: RequestLike) => Promise<any> | any;

export class ContextBuilder {
    private logger: ILogger;
    private ctx: any;
    private extenders: ContextExtender[] = [];
    private sharedContext: SharedContext | null = null;

    constructor(logger?: ILogger) {
        this.logger = logger ?? create('app-yaml-overwrites', 'context-builder.ts');
        this.ctx = {
            env: process.env,
            config: {},
            app: {},
            state: {},
            request: undefined,
            shared: null  // Will be set per resolution pass
        };
    }

    public withConfig(config: any): ContextBuilder {
        this.ctx.config = config;
        return this;
    }

    public withEnv(env: Record<string, string>): ContextBuilder {
        this.ctx.env = env;
        return this;
    }

    public withAppConfig(app: any): ContextBuilder {
        this.ctx.app = app;
        return this;
    }

    public withState(state: any): ContextBuilder {
        this.ctx.state = state;
        return this;
    }

    public withRequest(request: RequestLike): ContextBuilder {
        this.ctx.request = request;
        return this;
    }

    /**
     * Set shared context for resolution pass.
     * If not provided, creates a new SharedContext.
     *
     * The shared context allows computed functions to share state
     * during a single resolution pass (Option 5 pattern).
     *
     * @param shared - Optional existing SharedContext, or creates new one
     * @returns Self for chaining
     */
    public withSharedContext(shared?: SharedContext): ContextBuilder {
        this.sharedContext = shared ?? createSharedContext();
        return this;
    }

    public addExtender(extender: ContextExtender): ContextBuilder {
        this.extenders.push(extender);
        return this;
    }

    public async build(): Promise<any> {
        this.logger.debug('Building context', { keys: Object.keys(this.ctx) });

        let context = { ...this.ctx };

        // Always provide a shared context for computed functions
        context.shared = this.sharedContext ?? createSharedContext();

        for (const extender of this.extenders) {
            try {
                const partial = await extender(context, context.request);
                context = { ...context, ...partial };
            } catch (err) {
                this.logger.warn(`Context extender failed: ${err}`);
            }
        }

        return context;
    }

    /**
     * Static convenience method for building context.
     *
     * @param options - Context options (config, env, app, state, request, shared)
     * @param extenders - Optional list of context extenders
     * @param logger - Optional logger instance
     * @returns Built context dictionary with shared context included
     */
    public static async build(options: ContextOptions, extenders: ContextExtender[] = [], logger?: ILogger): Promise<any> {
        const builder = new ContextBuilder(logger);

        if (options.config) builder.withConfig(options.config);
        if (options.env) builder.withEnv(options.env);
        if (options.app) builder.withAppConfig(options.app);
        if (options.state) builder.withState(options.state);
        if (options.request) builder.withRequest(options.request);
        if (options.shared) builder.withSharedContext(options.shared);

        extenders.forEach(ext => builder.addExtender(ext));

        return builder.build();
    }
}

// Re-export SharedContext for convenience
export { SharedContext, createSharedContext } from './shared-context.js';

File: platform/polyglot/app_yaml_overwrites/py/src/app_yaml_overwrites/context_builder.py

import os
from typing import Dict, Any, Optional, List, Callable, Awaitable, Protocol, runtime_checkable
from .logger import ILogger, create
from .shared_context import SharedContext, create_shared_context


@runtime_checkable
class RequestLike(Protocol):
    """Generic request type for framework-agnostic usage."""
    headers: Optional[Dict[str, Any]]


ContextExtender = Callable[[Dict[str, Any], Optional[Any]], Awaitable[Dict[str, Any]]]

class ContextBuilder:
    def __init__(self, logger: ILogger = None):
        self._logger = logger or create("app_yaml_overwrites", "context_builder.py")
        self._ctx = {
             "env": dict(os.environ),
             "config": {},
             "app": {},
             "state": {},
             "request": None,
             "shared": None  # Will be set per resolution pass
        }
        self.extenders: List[ContextExtender] = []
        self._shared_context: Optional[SharedContext] = None

    def with_config(self, config: Dict[str, Any]) -> 'ContextBuilder':
        self._ctx["config"] = config
        return self

    def with_env(self, env: Dict[str, str]) -> 'ContextBuilder':
        self._ctx["env"] = env
        return self

    def with_app_config(self, app_config: Dict[str, Any]) -> 'ContextBuilder':
        self._ctx["app"] = app_config
        return self

    def with_state(self, state: Dict[str, Any]) -> 'ContextBuilder':
        self._ctx["state"] = state
        return self

    def with_request(self, request: Any) -> 'ContextBuilder':
        self._ctx["request"] = request
        return self

    def with_shared_context(self, shared: SharedContext = None) -> 'ContextBuilder':
        """
        Set shared context for resolution pass.
        If not provided, creates a new SharedContext.

        The shared context allows computed functions to share state
        during a single resolution pass (Option 5 pattern).

        Args:
            shared: Optional existing SharedContext, or creates new one

        Returns:
            Self for chaining
        """
        self._shared_context = shared or create_shared_context()
        return self

    def add_extender(self, extender: ContextExtender) -> 'ContextBuilder':
        self.extenders.append(extender)
        return self

    async def build(self) -> Dict[str, Any]:
        self._logger.debug("Building context", data={"keys": list(self._ctx.keys())})

        context = self._ctx.copy()

        # Always provide a shared context for computed functions
        context["shared"] = self._shared_context or create_shared_context()

        # Apply extenders
        if self.extenders:
            for extender in self.extenders:
                try:
                    partial = await extender(context, context["request"])
                    context.update(partial)
                except Exception as e:
                    self._logger.warn(f"Context extender failed: {e}")

        return context

    @staticmethod
    async def build_static(
        options: Dict[str, Any],
        extenders: List[ContextExtender] = None,
        logger: ILogger = None,
        shared: SharedContext = None
    ) -> Dict[str, Any]:
        """
        Static convenience method for backward compatibility / easy usage.

        Args:
            options: Context options (config, env, app, state, request)
            extenders: Optional list of context extenders
            logger: Optional logger instance
            shared: Optional shared context for computed functions

        Returns:
            Built context dictionary with shared context included
        """
        builder = ContextBuilder(logger)

        if "config" in options: builder.with_config(options["config"])
        if "env" in options: builder.with_env(options["env"])
        if "app" in options: builder.with_app_config(options["app"])
        if "state" in options: builder.with_state(options["state"])
        if "request" in options: builder.with_request(options["request"])
        if shared: builder.with_shared_context(shared)

        if extenders:
            for ext in extenders:
                builder.add_extender(ext)

        return await builder.build()
